File: yupana-jdbc/src/main/scala/org/yupana/jdbc/YupanaTcpClient.scala
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc

import org.yupana.api.query.SimpleResult
import org.yupana.api.types.DataType
import org.yupana.api.utils.CollectionUtils
import org.yupana.jdbc.YupanaConnection.QueryResult
import org.yupana.jdbc.YupanaTcpClient.Handler
import org.yupana.jdbc.build.BuildInfo
import org.yupana.protocol._

import java.net.{ InetSocketAddress, StandardSocketOptions }
import java.nio.ByteBuffer
import java.nio.channels.{ AsynchronousSocketChannel, CompletionHandler }
import java.util.concurrent.atomic.AtomicInteger
import java.util.logging.Logger
import java.util.{ Timer, TimerTask }
import scala.collection.mutable
import scala.concurrent.duration.Duration
import scala.concurrent.{ Await, ExecutionContext, Future, Promise }

class YupanaTcpClient(val host: String, val port: Int, batchSize: Int, user: String, password: String)(
    implicit ec: ExecutionContext
) extends AutoCloseable {

  private val logger = Logger.getLogger(classOf[YupanaTcpClient].getName)

  logger.info("New instance of YupanaTcpClient")

  private var channel: AsynchronousSocketChannel = _
  private var chanelReader: FramingChannelReader = _
  private val nextId: AtomicInteger = new AtomicInteger(0)
  private var iterators: Map[Int, ResultIterator] = Map.empty
  private val commandQueue: mutable.Queue[Handler[_]] = mutable.Queue.empty

  private val heartbeatTimer = new Timer()
  private val HEARTBEAT_INTERVAL = 30_000

  private var closed: Boolean = true

  private def ensureNotClosed(): Unit = {
    if (closed) throw new YupanaException("Connection is closed")
  }

  private def startHeartbeats(startTime: Long): Unit = {
    heartbeatTimer.schedule(
      new TimerTask {
        override def run(): Unit = sendHeartbeat(startTime)
      },
      HEARTBEAT_INTERVAL,
      HEARTBEAT_INTERVAL
    )
  }

  private def cancelHeartbeats(): Unit = {
    heartbeatTimer.cancel()
    heartbeatTimer.purge()
  }

  def prepareQuery(query: String, params: Map[Int, ParameterValue]): QueryResult = {
    val id = nextId.incrementAndGet()
    Await.result(execRequestQuery(id, SqlQuery(id, query, params)), Duration.Inf)
  }

  def batchQuery(query: String, params: Seq[Map[Int, ParameterValue]]): QueryResult = {
    val id = nextId.incrementAndGet()
    Await.result(execRequestQuery(id, BatchQuery(id, query, params)), Duration.Inf)
  }

  def connect(reqTime: Long): Unit = {
    logger.fine("Hello")

    if (channel == null || !channel.isOpen /* || !channel.isConnected*/ ) {
      logger.info(s"Connect to $host:$port")
      channel = AsynchronousSocketChannel.open()
      channel.setOption(StandardSocketOptions.SO_KEEPALIVE, java.lang.Boolean.TRUE)
      channel.connect(new InetSocketAddress(host, port)).get()

      chanelReader = new FramingChannelReader(channel, Frame.MAX_FRAME_SIZE + FramingChannelReader.PAYLOAD_OFFSET)
      closed = false
    }

    val cf = for {
      _ <- write(Hello(ProtocolVersion.value, BuildInfo.version, reqTime, Map.empty))
      _ <- waitHelloResponse(reqTime)
      cr <- waitFor(CredentialsRequest)
      _ <- sendCredentials(cr)
      _ <- waitFor(Authorized)
    } yield ()

    Await.result(cf, Duration.Inf)
    startHeartbeats(reqTime)
  }

  private def waitHelloResponse(reqTime: Long): Future[HelloResponse] = {
    waitFor(HelloResponse).flatMap { response =>
      if (response.protocolVersion != ProtocolVersion.value) {
        Future.failed(
          new YupanaException(
            error(
              s"Incompatible protocol versions: ${response.protocolVersion} on server and ${ProtocolVersion.value} in this driver"
            )
          )
        )
      } else if (response.reqTime != reqTime) {
        Future.failed(new YupanaException(error("got wrong hello response")))
      } else {
        Future.successful(response)
      }
    }
  }

  private def sendCredentials(cr: CredentialsRequest): Future[Unit] = {
    if (!cr.methods.contains(CredentialsRequest.METHOD_PLAIN)) {
      Future.failed(new YupanaException(error(s"All the auth methods ${cr.methods.mkString(", ")} are not supported")))
    } else {
      write(Credentials(CredentialsRequest.METHOD_PLAIN, user, password))
    }
  }

  private def sendHeartbeat(startTime: Long): Unit = {
    val time = (System.currentTimeMillis() - startTime) / 1000
    if (channel.isOpen) {
      Await.ready(write(Heartbeat(time.toInt)), Duration.Inf)
    }
  }

  private def runNext(): Future[Unit] = {
    commandQueue.headOption match {
      case Some(handler) =>
        for {
          _ <- write(handler.cmd)
          _ <- handler.execute()
          _ <- {
            commandQueue.synchronized {
              commandQueue.dequeue()
            }
            runNext()
          }
        } yield ()
      case None => Future.successful(())
    }
  }

  private def runCommand[T](cmd: Command[_], f: Promise[T] => Unit): Future[T] = {
    val p = Promise[T]()
    commandQueue.synchronized {
      commandQueue.enqueue(Handler(cmd, p, f))
    }
    runNext()

    p.future
  }

  private def readBatch(id: Int, read: Int): Future[Int] = {
    chanelReader.readFrame().flatMap { frame =>
      frame.frameType match {
        case Tags.RESULT_ROW =>
          val row = ResultRow.readFrame(frame)
          if (row.id == id) {
            iterators(id).addResult(row)
            if (read < batchSize) readBatch(id, read + 1) else Future.successful(read + 1)
          } else {
            Future.failed(new YupanaException(s"Unexpected row id ${row.id}"))
          }

        case Tags.RESULT_FOOTER =>
          iterators(id).setDone()
          Future.successful(read)

        case Tags.ERROR_MESSAGE =>
          val em = ErrorMessage.readFrame(frame)
          val msg = em.streamId.fold(s"A big problem here. ${em.message}") { i =>
            if (i == id) s"Unable to read batch. ${em.message}" else s"Unexpected row id $i in error. ${em.message}"
          }
          Future.failed(new YupanaException(msg))

        case x => Future.failed(new YupanaException(s"Unexpected response ${x.toChar} in Next handler"))
      }
    }
  }

  def acquireNext(id: Int): Unit = {
    assert(iterators.contains(id))
    val f = runCommand(
      Next(id, batchSize),
      (p: Promise[Unit]) => readBatch(id, 0).onComplete(x => p.complete(x.map(_ => ())))
    )
    Await.result(f, Duration.Inf)
  }

  def cancel(id: Int): Unit = {
    if (iterators.contains(id)) {
      val f = runCommand(
        Cancel(id),
        (p: Promise[Unit]) =>
          waitFor(Canceled).onComplete { cancelled =>
            cancelled.map { c =>
              assert(c.id == id)
              iterators.synchronized {
                iterators -= id
              }
            }
            p.complete(cancelled.map(_ => ()))
          }
      )

      Await.result(f, Duration.Inf)
    }
  }

  private def waitFor[T <: Message[T]](helper: MessageHelper[T])(
      implicit ec: ExecutionContext
  ): Future[T] = {
    chanelReader.readFrame().flatMap { frame =>
      frame.frameType match {
        case helper.tag => Future.successful(helper.readFrame[ByteBuffer](frame))
        case Tags.ERROR_MESSAGE =>
          val msg = ErrorMessage.readFrame(frame).message
          Future.failed(new YupanaException(error(s"Got error response on '${helper.tag.toChar}', '$msg'")))

        case x =>
          Future.failed(
            new YupanaException(error(s"Unexpected response '${x.toChar}' while waiting for '${helper.tag.toChar}'"))
          )
      }
    }
  }

  private def execRequestQuery(id: Int, command: Command[_]): Future[QueryResult] = {
    logger.fine(s"Exec request query $command")
    ensureNotClosed()
    runCommand(
      command,
      (p: Promise[ResultIterator]) =>
        waitFor(ResultHeader).onComplete { header =>
          p.complete(header.map(h => {
            iterators.synchronized {
              val r = new ResultIterator(h, this)
              iterators += id -> r
              r
            }
          }))
        }
    ).map(it => extractProtoResult(id, it))
  }

  private def write(request: Command[_]): Future[Unit] = {
    logger.fine(s"Writing command ${request.helper.tag.toChar}")
    val f = request.toFrame
    val bb = ByteBuffer.allocate(f.payload.length + 4 + 1)
    bb.put(f.frameType)
    bb.putInt(f.payload.length)
    bb.put(f.payload)
    bb.flip()

    JdbcUtils.wrapHandler[Unit](
      new CompletionHandler[Integer, Promise[Unit]] {
        override def completed(result: Integer, p: Promise[Unit]): Unit = p.success(())
        override def failed(exc: Throwable, p: Promise[Unit]): Unit = p.failure(exc)
      },
      (p, h) => channel.write(bb, p, h)
    )
  }

  private def error(e: String): String = {
    logger.warning(s"Got error message: $e")
    e
  }

  override def close(): Unit = {
    logger.fine("Close connection")
    closed = true
    cancelHeartbeats()
    Await.ready(write(Quit()), Duration.Inf)
    channel.close()
  }

  private def extractProtoResult(id: Int, res: ResultIterator): QueryResult = {
    val header = res.header
    val names = header.fields.map(_.name)
    val dataTypes = CollectionUtils.collectErrors(header.fields.map { resultField =>
      DataType.bySqlName(resultField.typeName).toRight(s"Unknown type ${resultField.typeName}")
    }) match {
      case Right(types) => types
      case Left(err)    => throw new IllegalArgumentException(s"Cannot read data: $err")
    }

    val values = res.map { row =>
      dataTypes
        .zip(row.values)
        .map {
          case (rt, bytes) =>
            if (bytes.isEmpty) {
              null
            } else {
              rt.storable.read(bytes)
            }
        }
        .toArray
    }

    QueryResult(id, SimpleResult(header.tableName, names, dataTypes, values))
  }
}

object YupanaTcpClient {
  private case class Handler[R](cmd: Command[_], promise: Promise[R], f: Promise[R] => Unit) {
    def execute(): Future[R] = {
      f(promise)
      promise.future
    }
  }
}

File: yupana-jdbc/src/main/java/org/yupana/jdbc/YupanaTcpClient.java
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ProtocolException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.SocketChannel;
import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import com.google.protobuf.InvalidProtocolBufferException;
import org.yupana.Proto.*;
import org.yupana.proto.util.ProtocolVersion;

import org.yupana.jdbc.build.BuildInfo;

public class YupanaTcpClient implements AutoCloseable {

  private final String host;
  private final int port;
  private SocketChannel channel;

  private static final int CHUNK_SIZE = 1024 * 100;
  private static final Logger logger = Logger.getLogger(YupanaTcpClient.class.getName());

  public YupanaTcpClient(String host, int port) {
    this.host = host;
    this.port = port;
  }


  private void ensureConnected() throws IOException {
    if (channel == null || !channel.isConnected()) {
      channel = SocketChannel.open();
      channel.configureBlocking(true);
      channel.connect(new InetSocketAddress(host, port));
    }
  }

  public Iterator<Object> query(String query, Map<Integer, ParameterValue> params) throws IOException {
    Request request = createProtoQuery(query, params);
    return execRequestQuery(request);
  }

  public Iterator<Object> batchQuery(String query, Collection<Map<Integer, ParameterValue>> params) throws IOException {
    Request request = creteProtoBatchQuery(query, params);
    return execRequestQuery(request);
  }

  Optional<Version> ping(Long reqTime) throws IOException {
    Request request = createProtoPing(reqTime);
    Pong pong = execPing(request);

    if (pong.getReqTime() != reqTime) {
      throw new ProtocolException("got wrong ping response");
    }
    channel.close();

    return Optional.ofNullable(pong.getVersion());
  }

  private Pong execPing(Request request) throws IOException {
    ensureConnected();
    sendRequest(request);
    Response response = fetchResponse();

    switch (response.getRespCase()) {
      case PONG:
        Pong pong = response.getPong();
        if (pong.getVersion().getProtocol() != ProtocolVersion.value)
          throw new ProtocolException(
                  String.format("Incompatible protocol versions: %d on server and %d in this driver",
                          pong.getVersion().getProtocol(),
                          Protocolversion.value)
          );
        return pong;

      case ERROR:
        throw new IOException(error(String.format("Got error response on ping, '%s'", response.getError())));

      default:
        throw new IOException(error("Unexpected response on ping"));
    }
  }

  private Iterator<Object> execRequestQuery( Request request) throws IOException {
    ensureConnected();
    sendRequest(request);
    FramingChannelIterator it = new FramingChannelIterator(channel, CHUNK_SIZE + 4);

    try {
      Stream<Response> resps = StreamSupport.stream(Spliterators.spliteratorUnknownSize(it, Spliterator.ORDERED), false)
              .map(Response::parseFrom);
    } catch (InvalidProtocolBufferException e) {
      e.printStackTrace();
    }

    val header = it.map(resp => handleResultHeader(resp)).find(_.isDefined).flatten

    header match {
      case Some(Right(h)) =>
        val r = resultIterator(it)
        extractProtoResult(h, r)

      case Some(Left(e)) =>
        channel.close()
        throw new IllegalArgumentException(e)

      case None =>
        channel.close();
        throw new IllegalArgumentException(error("Result not received"));
    }
  }

  private void sendRequest( Request request) throws IOException {
    try {
      channel.write(createChunks(request.toByteArray()));
    } catch (IOException io) {
      logger.warning(String.format("Caught %s while trying to write to channel, let's retry", io.getMessage()));
      try {
        Thread.sleep(1000);
      } catch (InterruptedException e) {
        throw new IOException(e);
      }
      ensureConnected();
      channel.write(createChunks(request.toByteArray()));
    }
  }

  private ByteBuffer[] createChunks(byte [] data) {
    int last = data.length % CHUNK_SIZE;
    int chunks = data.length / CHUNK_SIZE + (last > 0 ? 1 : 0);
    ByteBuffer[] result  = new ByteBuffer[chunks];
    for (int i = 0; i < chunks - (last > 0 ? 1: 0); i++) {
      result[i] = ByteBuffer.allocate(CHUNK_SIZE + 4).order(ByteOrder.BIG_ENDIAN);
      result[i].putInt(CHUNK_SIZE);
      result[i].put(data, i * CHUNK_SIZE, CHUNK_SIZE);
    }

    if (last > 0) {
      result[chunks - 1] = ByteBuffer.allocate(last + 4).order(ByteOrder.BIG_ENDIAN);
      result[chunks - 1].putInt(last);
      result[chunks - 1].put(data, (chunks - 1) * CHUNK_SIZE, last);
    }

    return result;
  }

  private Response fetchResponse() throws IOException {
    ByteBuffer bb = ByteBuffer.allocate(CHUNK_SIZE + 4).order(ByteOrder.BIG_ENDIAN);
    int bytesRead = channel.read(bb);
    if (bytesRead < 0) throw new IOException("Broken pipe");
    else if (bytesRead < 4) throw new IOException("Invalid server response");
    bb.flip();
    int chunkSize = bb.getInt();
    byte[] bytes = new byte[chunkSize];
    bb.get(bytes);
    return Response.parseFrom(bytes);
  }

  private Optional<ResultHeader> handleResultHeader(Response response) throws IOException {
    switch (response.getRespCase()) {
      case RESULTHEADER:
        ResultHeader h = response.getResultHeader();
        logger.info("Received result header " + h);
        return Optional.of(h);

      case RESULT:
        throw new IOException(error("Data chunk received before header"));

      case PONG:
        throw new IOException(error("Unexpected TspPong response"));

      case HEARTBEAT:
        heartbeat(response.getHeartbeat());
        return Optional.empty();

      case ERROR:
        channel.close();
        throw new IOException(error(response.getError()));

      case RESULTSTATISTICS:
        throw new IOException(error("Unexpected ResultStatistics response"));

      case RESP_NOT_SET:
        return Optional.empty();
    }
  }

  private String error(String e) {
    logger.warning(String.format("Got error message: %s", e));
    return e;
  }

  private void heartbeat(String time) {
    logger.info(String.format("Heartbeat(%s)", time));
  }

  private Iterator<ResultChunk> resultIterator(Iterator<Response> responses) throws IOException {
    return new Iterator<ResultChunk>() {

      {
        readNext();
      }

      @Override
      public boolean hasNext() {
        return responses.hasNext() && statistics == null && errorMessage == null;
      }

      @Override
      public ResultChunk next() {
        ResultChunk result = current;
        readNext();
        return result;
      }

      private ResultStatistics statistics = null;
      private ResultChunk current = null;
      private String errorMessage = null;


      private void readNext() throws IOException {
        current = null;
        do {
          Response next = responses.next();
          switch (next.getRespCase()) {
            case RESULT:
              current = next.getResult();
              break;

            case RESULTHEADER:
              errorMessage = error("Duplicate header received");
              break;

            case PONG:
              errorMessage = error("Unexpected Pong response");
              break;

            case HEARTBEAT:
              heartbeat(next.getHeartbeat());
              break;

            case ERROR:
              errorMessage = error(next.getError());
              break;

            case RESULTSTATISTICS:
              logger.fine(String.format("Got statistics %s", next.getResultStatistics().toString()));
              statistics = next.getResultStatistics();
              break;

            case RESP_NOT_SET:
              break;
          }
        } while (current == null && statistics == null && errorMessage == null && responses.hasNext());

        if (statistics != null || errorMessage != null) {
          channel.close();
          if (errorMessage != null) {
            throw new IllegalArgumentException(errorMessage);
          }
        }

        if (!responses.hasNext() && statistics == null) {
          channel.close();
          throw new IllegalArgumentException("Unexpected end of response");
        }
      }
    };
  }

  @Override
  public void close() throws IOException {
    channel.close();
  }

  private Request createProtoPing(Long reqTime) {
    return Request.newBuilder()
            .setPing(Ping.newBuilder()
                    .setReqTime(reqTime)
                    .setVersion(Version.newBuilder()
                            .setProtocol(ProtocolVersion.value)
                            .setMajor(BuildInfo.majorVersion)
                            .setMinor(BuildInfo.minorVersion)
                            .setVersion(BuildInfo.version)
                    )
            ).build();
  }

  private Iterator<Object> extractProtoResult(ResultHeader header, Iterator<ResultChunk> res) {
    List<String> names = header.getFieldsList().stream().map(ResultField::getName).collect(Collectors.toList());
    List<String> errors = new ArrayList<>();
    List<DataType> dataTypes = new ArrayList<>();

    header.getFieldsList().forEach(x -> {
      if () {
        dataTypes.add();
      } else {
        errors.add(String.format("Unknown type %s", x.getType()));
      }
    });

    if (!errors.isEmpty()) {
      throw new IllegalArgumentException(String.format("Cannot read data: %s", String.join(", ", errors)));
    }

    val values = res.map { row =>
      dataTypes
        .zip(row.values)
        .map {
          case (rt, bytes) =>
            if (bytes.isEmpty) {
              null
            } else {
              rt.storable.read(bytes.toByteArray)
            }
        }
        .toArray
    }

    return new SimpleResult(Optional.ofNullable(header.getTableName()).orElse("TABLE"), names, dataTypes, values);
  }

  private Request createProtoQuery(String query, Map<Integer, ParameterValue> params) {
    SqlQuery.Builder sq = SqlQuery.newBuilder().setSql(query);
    params.forEach(sq::setParameters);


    return Request.newBuilder().setSqlQuery(sq).build();
  }

  private Request creteProtoBatchQuery(String query, Collection<Map<Integer, ParameterValue>> params) {

    BatchSqlQuery.Builder sbq = BatchSqlQuery.newBuilder().setSql(query);

    params.forEach(batch -> {
      ParameterValues.Builder values = ParameterValues.newBuilder();
      batch.forEach(values::setParameters);
      sbq.addBatch(values);
    });
    return Request.newBuilder().setBatchSqlQuery(sbq).build();
  }
}
